### PR TITLE
use_state fold updater functions

### PR DESCRIPTION
**Pull request by jamesdbrock**
The `use_state` setter can take either a state value, or a state updater function, like in React.

image

### CODE REVIEW

File: edifice/engine.py [hunk 1]
```diff
@@ -621,6 +624,25 @@ def _render(self, component: Element, render_context: _RenderContext):
...
+
+        # Before the render, reduce the _hook_state updaters.
+        # We can't do this after the render, because there may have been state
+        # updates from event handlers.
+        for hooks in self._hook_state.values():
+            for hook in hooks:
+                state0 = hook.state
+                try:
+                    for updater in hook.updaters:
+                        if callable(updater):
```
Comment (jamesdbrock) [inline edifice/engine.py:636]: Oh no I just realized that this won't work because the user might be storing a callable in the state.

Comment (considerate) [reply edifice/engine.py:636]: I think that's fine. We could document that it needs to wrapped in a data structure like a tuple if you need to store a function.

File: edifice/hooks.py [hunk 2]
```diff
@@ -21,27 +33,49 @@ def Stateful(self):
 
             return Label(text=str(x))
 
+    If an **updater function** is passed to the **setter function**, then at the end of
+    the render the state will be modified by calling all of the
+    **updater functions** in this order in which they were passed.
+    An **updater function** is a function from state to state.
```
Comment (considerate) [inline edifice/hooks.py:39]:
```suggestion
    An **updater function** is a function from the previous state to the new state.
```

**Comment by considerate** [comment]
Do we want action dispatched folds as well (that would be another hook)?

https://react.dev/reference/react/useReducer

File: edifice/engine.py (hunk 1 above)
Comment (jamesdbrock) [reply edifice/engine.py:636]: Good idea.

File: edifice/hooks.py (hunk 2 above)
Comment (jamesdbrock) [reply edifice/hooks.py]: Done thanks

**Comment by considerate** [comment]
> Heck yes but it's low on my priority list. Can we write use_reducer in terms of use_state?

Yes, those should be isomorphic.

**Pull request merged**